app/models/conversation.rb

```diff
@@ -13,6 +13,10 @@ class Conversation < ApplicationRecord
 
   scope :blocked, -> { where.not(developer_blocked_at: nil).or(Conversation.where.not(business_blocked_at: nil)) }
   scope :visible, -> { where(developer_blocked_at: nil, business_blocked_at: nil) }
+  scope :archived_by_business, -> { where.not(business_archived_at: nil).order(business_archived_at: :desc) }
+  scope :unarchived_by_business, -> { where(business_archived_at: nil).order(updated_at: :desc) }
+  scope :archived_by_developer, -> { where.not(developer_archived_at: nil).order(developer_archived_at: :desc) }
+  scope :unarchived_by_developer, -> { where(developer_archived_at: nil).order(updated_at: :desc) }
```

Comment:
I think that these scopes aren't the right abstraction because they are requiring you to use `#send`. I think we might be able to consolidate to 2 scopes and pass in the user - `archived_by_user` and `unarchived_by_user`, maybe?

Reply:
Done ✅

Reply:
Here's the alternatives to those methods:
```
  scope :archived_by_user, -> (dev_or_business) { where("#{dev_or_business} IS NOT NULL").order("#{dev_or_business} DESC") }
  scope :unarchived_by_user, -> (dev_or_business) { where("#{dev_or_business} IS NULL").order("#{dev_or_business} DESC") }
```

Reply:
That string interpolation directly in a query scares me. Is there no way to add this to the User model so we have the scope already?

Reply:
Exactly! I knew you would say this, so here's a different approach I came up with:

In the `Conversation` Model:

```
  scope :archived_by_user, (lambda do |user|
    user.business ? where.not(business_archived_at: nil) : where.not(developer_archived_at: nil)
  end)

  scope :unarchived_by_user, (lambda do |user|
    user.business ? where(business_archived_at: nil) : where(developer_archived_at: nil)
  end)
```
Instead of passing the right attribute as an argument and then using string interpolation to extract results, I thought why not just pass in the user as an argument to it. This avoids string interpolation in the scope, removes the extra logic from both the controller, and makes code more readable.

Now, of course, this doesn't solve the issue of a user having both `developer` & `business` types but we can extend that logic into this in the future.

Reply:
To be honest, when I read the issue of this PR I thought this task wouldn't take more than a couple of hours, but boy I was wrong! 😄

Reply:
It shouldn't take more than a few hours. The whole developer vs. business thing is complicating the issue a ton. Which is no fault of yours! But is definitely calling out a bigger problem in the data modeling.

Reply:
Interestingly enough, there is a single person with both developer and business conversations. Me :)

Another approach to all of this would be to have an `Archive` model joins between conversations and business/developer. I wonder if that would simplify the logic of these queries?

If not, or we don't want to explore that path right now, I'm happy to hold off on this issue until #237 is finally completed. What do you think?

Reply:
> Interestingly enough, there is a single person with both developer and business conversations. Me :)

Haha! But I believe that in the future maybe you have users who are both developers and businesses. And for that I think I have the solution and it would be just an extra check to the existing scope that I showed:

```
  scope :unarchived_by_user, (lambda do |user|
    if user.developer && user.business
      where(business_archived_at: nil, developer_archived_at: nil).order(updated_at: :desc)
    elsif user.business
      where(business_archived_at: nil).order(updated_at: :desc)
    elsif user.developer
      where(developer_archived_at: nil).order(updated_at: :desc)
    end
  end)
```
And we can share this same logic to the `archived_by_user` scope.

I will see how I can refactor this but wouldn't this solve our query, at least for now? What do you think?

Reply:
I don't think that logic will work the way we want to.

Just because a user has a business account doesn't mean they are the business associated with every conversations. Meaning, I can have a business account but be involved with a conversation as a developer. If _that_ business archives the conversation then it would be considered archived by your code.

Reply:
Well! How about this?
```
  scope :archived_by_business, -> { where.not(business_archived_at: nil) }
  scope :unarchived_by_business, -> { where(business_archived_at: nil) }
  scope :archived_by_developer, -> { where.not(developer_archived_at: nil) }
  scope :unarchived_by_developer, -> { where(developer_archived_at: nil) }

  scope :unarchived_by_user, (lambda do |user|
    if user.developer && user.business
      unarchived_by_business.merge(unarchived_by_developer)
    elsif user.business
      unarchived_by_business
    elsif user.developer
      unarchived_by_developer
    end
  end)
```

Reply:
Again, I don't think that logic will work. We need to check _each_ conversation to see if the user is on the business side or developer side to see which "archived at" timestamp is set or not.

I think that this comment is as close as we've gotten to getting that to work. Have you explored this?

https://github.com/joemasilotti/railsdevs.com/pull/813#discussion_r1142159773

Reply:
In the above query, as we can see we're giving or passing the `current_user` twice, in order to refactor that, and as you suggested, I have implemented that logic in the `User` model like this:
```
# app/models/user.rb

  def archived_conversations
    if business && developer
      conversations.where.not(business_archived_at: nil).merge(where.not(developer_archived_at: nil))
    elsif business
      conversations.where.not(business_archived_at: nil)
    elsif developer
      conversations.where.not(developer_archived_at: nil)
    end
  end

  def unarchived_conversations
    if developer && business
      conversations.where(business_archived_at: nil).merge(where(developer_archived_at: nil))
    elsif business
      conversations.where(business_archived_at: nil)
    elsif developer
      conversations.where(developer_archived_at: nil)
    end
  end
```

Now in our controllers we can call for those conversations like this:
```
current_user.unarchived_conversations
```
OR
```
current_user.archived_conversations
```

Reply:
I think you correctly understand the goal but are missing an annoyingly tiny nuance that makes the logic need some fine tuning.

Let's take a step back and dive into how conversations are queried. Here's what currently exists on `main` in the `User` model.

```ruby
# app/models/user.rb

has_many :conversations, ->(user) {
  unscope(where: :user_id)
    .left_joins(:business, :developer)
    .where("businesses.user_id = ? OR developers.user_id = ?", user.id, user.id)
    .visible
}
```

Note that we have to match an `OR` clause to make sure this user's developer _or_ business profile is part of the conversation.

Now, in this PR, a conversation can be archived by a business and/or a developer. But it is only archived to the user if they are _that_ business or developer.

So we need to make sure that "archived" shows conversations the user archived _as a business_ or _as a developer_. Said another way, if I message you as a business you are the developer. If I archive the conversation it is _not_ archived for you.

Let's assume you also have a business profile. Then _only_ checking for the presence of `business_archived_at` means it would show as archived for you. Which is wrong. I archived it!

Which is where this code (from the comment above) comes into place. It lines up the conversations a user has as a developer with the developer archive column and the same for business.

```ruby
has_many :conversations, ->(user) {
  conversations_as_developer.merge(conversations_as_business)
}

has_many :archived_conversations, ->(user) {
  archived_conversations_as_developer.merge(archived_conversations_as_business)
}

private

has_many :conversations_as_developer, ->(user) {
  unscope(where: :user_id)
    .left_joins(:developer)
    .where(developers: {user_id: user.id})
    .where(developer_archived_at: nil)
    .visible
}

has_many :conversations_as_business, ->(user) {
  unscope(where: :user_id)
    .left_joins(:business)
    .where(businesses: {user_id: user.id})
    .where(business_archived_at: nil)
    .visible
}

has_many :archived_conversations_as_developer, ->(user) {
  unscope(where: :user_id)
    .left_joins(:developer)
    .where(developers: {user_id: user.id})
    .where.not(developer_archived_at: nil)
    .visible
}
has_many :archived_conversations_as_business, ->(user) {
  unscope(where: :user_id)
    .left_joins(:business)
    .where(businesses: {user_id: user.id})
    .where.not(business_archived_at: nil)
    .visible
}
```

Does that make more sense?

Reply:
Wow! That's a detailed explanation! I never thought of the scenario where a business could also message other businesses and this is where the logic I created breaks, right?

Sorry I wasn't able to reply back to you as our holy month of Ramadan has begun so I will be much busier than usual but I will make sure to keep in touch with you every 3-4 days. 

I will be experimenting with the logic you have given soon and revert to you if I have any doubts or queries. I was also thinking maybe we can create a different model as you suggested earlier to keep track of which `user_id` has archived and which hasn't. Your thoughts?

Reply:
> I never thought of the scenario where a business could also message other businesses and this is where the logic I created breaks, right?

Not exactly, only a developer and a business can message each other. But a user might have a business profile but be acting as a developer in a specific conversations. I hope that distinction makes sense!

You know... I wonder if a `Conversations::Block` record would simplify a ton of this logic. Do you want to explore that path or continue trucking with what you have?

And have a generous Ramadan! Let me know if you'd like to pause on this or pass it off. I'm happy with whatever you choose but I need to make sure work continues as other folks will want to pick up stale issues.

Reply:
> A user can be a business type but a developer to a conversation.

That makes so much sense now to me. I am open to any ideas you have because I don't think my logic could work in this scenario & you knew it but I couldn't understand the root of it.

Also I want to complete this PR as quick as possible. I believe this has taken much more time than expected. So yes, do let me know what you have in mind for such a problem. 

Thanks for the wishes! Happy weekend! 🙂

Reply:
@kaspth proposed a refactor a while back that converts "blocking" conversations to its own model. We could adopt that same approach for "archiving" conversations.

https://github.com/joemasilotti/railsdevs.com/pull/455

What I like about this approach is that it is per _user_ not developer/business. Which should make a lot of querying much simpler!

Reply:
> @kaspth proposed a refactor a while back that converts "blocking" conversations to its own model. We could adopt that same approach for "archiving" conversations.
> 
> #455 

I did read it but I couldn't understand how could I use that logic here.

So I read your previous replies to come up with this solution:

*`conversations_controller.rb`*

```
  def index
    @conversations = current_user.conversations
    @archived_any = current_user.archived_conversations.exists?
  end

```

*`app/models/user.rb`*

```
  def archived_conversations
    Conversation.archived_conversations_as_developer(self).union(Conversation.archived_conversations_as_business(self))
  end

  def conversations
    Conversation.conversations_as_developer(self).union(Conversation.conversations_as_business(self))
  end
```

*`app/models/coversation.rb`*

```
  scope :conversations_as_developer, ->(user) {
    unscope(where: :user_id)
    .left_joins(:developer)
    .where(developers: {user_id: user.id})
    .where(developer_archived_at: nil)
    .visible
  }

  scope :conversations_as_business, ->(user) {
    unscope(where: :user_id)
    .left_joins(:business)
    .where(businesses: {user_id: user.id})
    .where(business_archived_at: nil)
    .visible
  }

  scope :archived_conversations_as_developer, ->(user) {
    unscope(where: :user_id)
    .left_joins(:developer)
    .where(developers: {user_id: user.id})
    .where.not(developer_archived_at: nil)
    .visible
  }

  scope :archived_conversations_as_business, ->(user) {
    unscope(where: :user_id)
    .left_joins(:business)
    .where(businesses: {user_id: user.id})
    .where.not(business_archived_at: nil)
    .visible
  }
```

You must be wondering where the `union` method comes from. That comes from the gem I installed [`active_record_union`](https://github.com/brianhempel/active_record_union). The `merge` method doesn't return the results we're looking for.

Thoughts?

Reply:
heyo, chiming in since I was mentioned. I still find the business/developer split, so I just wanted to show a sketch that tries to blend the two — and then how I'd do archiving off of that established relationship. It probably doesn't work yet, since the domain modeling here between conversations and whose the sender/recipient is still murky to me. But nonetheless, here's a sketch:

```ruby
class Account < ApplicationRecord
  delegated_type :record, types: %w[ Business Developer ]

  has_many :memberships, dependent: :destroy
  has_many :users, through: :memberships
end

class User::Membership < ApplicationRecord
  belongs_to :account
  belongs_to :user
end

class User < ApplicationRecord
  has_many :memberships
  has_many :accounts, through: :memberships # So businesses could have multiple users attached
  delegate :businesses, :developers, to: :accounts

  has_many :conversations, through: :accounts
  has_many :archivings
end

class Conversation < ApplicationRecord
  belongs_to :account
  has_many :archivings, dependent: :destroy

  scope :archived_with, ->(user) { joins(:archivings).where(archivings: { account: user.accounts }) }
  scope :visible_to,    ->(user) { left_outer_joins(:archivings).where(archivings: { account: user.accounts }) }

  def business?(user)
    user.businesses.include? account.record
  end

  def developer?(user)
    user.developers.include? account.record
  end
end

class Conversation::Archiving < ApplicationRecord
  belongs_to :conversation
  belongs_to :account
  belongs_to :user
end

class Conversations::ArchivingsController < ApplicationController
  before_action :authenticate_user!

  def index
    @conversations = Conversation.archived_with(current_user)
  end

  def create
    conversation.archivings.create!(account:, user: current_user)
    redirect_to conversations_url, notice: "Conversation archived"
  end

  def destroy
    conversation.archivings.destroy_by!(account:)
    redirect_to conversations_url, notice: "Conversation unarchived"
  end

  private

  def conversation
    @conversation ||= Conversation.find(params[:id])
  end

  def account
    @account ||= current_user.accounts.find(params[:account_id])
  end
end

class ConversationsController < ApplicationController
  before_action :authenticate_user!

  def index
    @conversations = Conversation.visible_to(current_user)
    @archived_any = current_user.archivings.exists?
  end

  def show
    authorize @conversation = Conversation.find(params[:id])

    @conversation.mark_notifications_as_read(current_user)

    @message = @conversation.messages.new
  end
end

# Potential alternative sketch where an archiving and a block are really considered the same type of model,
# because we're connecting an action on a conversation between a user and an account.
# Didn't get far with this.
class Conversation::Involvement < ApplicationRecord
  belongs_to :conversation
  belongs_to :account, default: -> { conversation&.account }
  belongs_to :user
  
  enum kind: %i[ block archive ].index_by(&:itself)
end

class Conversations::InvolvementsController < ApplicationController
  def create
  end
end
```

Reply:
Thanks @kaspth! Once RailsDevs has proper accounts (instead of only users) this looks pretty ideal to me. Unfortunately, there is a lot more work to get us there right now.

This remodeling also starts to immediately open up interesting product questions. Does a user or account archive a conversation? What if a team wants to share a workflow and message developers together? What about blocks? Who _owns_ one of those? Interesting stuff!

@JuzerShakir, two questions:

1. What is the difference between a union and a merge? Could an `or` work just the same? I'm not sure I understand the nuances here.
2. I'm starting to think we might want to postpone this PR until #410 is finally live. Someone is starting to work on that now but I don't have a good timeline on when it might be done.

What do you think?

Reply:
Answering  your questions:

**Could an `or` work the same?**
Yes! It does. `union` & `or` are quite similar where both are used to combine the results of multiple queries. I used the `or` method in our `scopes` but since it performs a `left_outer_join`, it shows the following error:
![Screenshot from 2023-03-26 11-31-37](https://user-images.githubusercontent.com/10782481/227758316-95542438-b396-4515-9412-d3d70f083490.png)

-----

**What is the difference between a union and a merge?**
Applying `merge` doesn't return any errors but it doesn't return the results we are looking for because when we call `merge` on a relation object as an argument, **ActiveRecord will combine the conditions of both relations into a single SQL query**. For example, consider the following two relation objects:
```
relation1 = User.where(age: 20)
relation2 = User.where(city: 'New York')
```
If we want to combine these two relation objects into a single relation object that contains all the users who are 20 years old and live in New York, we can use the `merge` method. But in our case, we don't want to combine the conditions but evaluate and return the queries separately and then return a single `ActiveRecord::Relation` object containing the results of both queries.

So in order to combine the results of those queries into a single result I could use `+` method which would give results that we expect but it would return an `Array` object instead of an `ActiveRecord::Relation` where we wouldn't be able to use any ActiveRecord methods or make SQL queries.

And that's where the `union` method comes in. The `union` method is used to combine the results of two or more queries into a single relation object with duplicates removed. The columns in each query must be in the same order and have the same data types for the `union` to work correctly. And the `active_record_union` gem enables us to use the `union` method to return the results we're looking for.

I hope this clears your doubt and explains my reasoning well for using the `union` query.

Reply:
That does - thanks for the detailed explanation!

What are your thoughts on the second question?

Reply:
Yeah sorry @joemasilotti, about that, I think you should do what you think is best. But I have a question, if this PR gets postponed will I be assigned to other issues that interest me for railsdev.com?

Reply:
OK, cool. Let's postpone this issue for now. I'll mark it blocked until #843 is done.

And yes, 100%. You are free to pick up any issue that is available. I'm also happy to pay you the bounty for this one since you put in so much work already. [Send me an email](mailto:joe@railsdevs.com) to coordinate.

Reply:
Thanks, a lot @joemasilotti ❤️